roles/tests-integration/tests/common/proxy/mod.rs

```diff
@@ -0,0 +1,489 @@
+            DeclareMiningJobError, DeclareMiningJobSuccess, IdentifyTransactions,
+            IdentifyTransactionsSuccess, ProvideMissingTransactions,
+            ProvideMissingTransactionsSuccess, SubmitSolution,
+        },
+        PoolMessages,
+        TemplateDistribution::{self, CoinbaseOutputDataSize},
+    },
+    utils::Mutex,
+};
+use tokio::{
+    net::{TcpListener, TcpStream},
+    select,
+    sync::mpsc::{channel, Receiver, Sender},
+};
+
+use codec_sv2::framing_sv2::framing::Frame as EitherFrame;
+
+type MessageFrame = StandardEitherFrame<PoolMessages<'static>>;
+
+#[derive(Debug, Clone)]
+pub struct SV2Proxy {
+    aggregator: MessagesAggregator,
+}
+
+impl SV2Proxy {
+    pub fn new() -> Self {
+        let aggregator = MessagesAggregator::new();
+        Self { aggregator }
+    }
+
+    pub async fn start(&self, upstream: SocketAddr, downstream: SocketAddr) {
+        let mut builder = ProxyBuilder::new();
+        builder
+            .try_add_client(Self::wait_for_client(downstream).await)
+            .await
+            .expect("Impossible to add client to the ProxyBuilder")
+            .try_add_server(
+                TcpStream::connect(upstream)
+                    .await
+                    .expect("Impossible to connect to server"),
+            )
+            .await
+            .expect("Impossible to add server to the ProxyBuilder");
+        self.listen_for_all_messages(&mut builder);
+        let err = builder
+            .start()
+            .await
+            .expect("Impossible to build the Client");
+        eprintln!("{err:?}");
+    }
+
+    pub fn contains(&self, message: ExpectMessage, role: Role) -> bool {
+        self.aggregator.contains(message, role)
+    }
+
+    fn listen_for_all_messages(&self, builder: &mut ProxyBuilder) {
+        for n in 0..44 {
+            let mut server_recv = builder.add_handler(Role::Server, n);
+            let mut client_recv = builder.add_handler(Role::Client, n);
+            let aggregator = self.aggregator.clone();
+            tokio::spawn(async move {
+                tokio::select! {
+                    m = server_recv.recv() => {
+                        aggregator.add_message(m.unwrap(), Role::Server);
+                    }
+                    m = client_recv.recv() => {
+                        aggregator.add_message(m.unwrap(), Role::Client);
+                    }
+                }
+            });
+        }
+    }
+
+    async fn wait_for_client(client: SocketAddr) -> TcpStream {
+        let listner = TcpListener::bind(client)
+            .await
+            .expect("Impossible to listen on given address");
+        if let Ok((stream, _)) = listner.accept().await {
+            stream
+        } else {
+            panic!("Impossible to accept dowsntream connetion")
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub enum ExpectMessage {
+    SetupConnection,
+    SetupConnectionSuccess,
+    SetupConnectionError,
+}
```

Comment:
what is the purpose of this struct?

why can't we simply assert against the messages that were already declared under `protocols/v2/subprotocols` (and re-exported via `roles_logic_sv2`)?

Reply:
`roles_logic_sv2::parsers::PoolMessages` already covers all possible messages

tbh I don't really like this name because it implies this should only be used on pools, but that is a different discussion... maybe we can improve this in the future, but for now, I think we can simply use `roles_logic_sv2::parsers::PoolMessages` instead of `ExpectMessage`

unless I'm missing something?

Reply:
also, `ExpectMessage` doesn't allow us to assert against message flags or parameters

Reply:
Yea the problem with using `PoolMessages` for assertion is that it does not implement `PartialEq`, neither all of it nested(..and nested) properties. If we think this is a hard requirement we can go and add `PartialEq` all over those structs/enums. In regards for asserting flags or other params, that is possible even with `ExpectMessage`, for example we can change `ExpectMessage::SetupConnect` to `ExpectMessage::SetupConnect(flag1)` or whatever we think is worth asserting in `SetupConnection`.

Reply:
> In regards for asserting flags or other params, that is possible even with `ExpectMessage`, for example we can change `ExpectMessage::SetupConnect` to `ExpectMessage::SetupConnect(flag1)` or whatever we think is worth asserting in `SetupConnection`.

in the end of the day, wouldn't `ExpectMessage` end up being a massive enum with a bunch of redundant info?

I think it's fine that we do a separate PR introducing `PartialEq` to `roles_logic_sv2::parsers::{CommonMessages, Mining, JobDeclaration, TemplateDistribution}`, that should be a very simple and straightforward change

> neither all of it nested(..and nested) properties.

from a quick inspection, it seems all the nested structs coming from `protocols/v2/subprotocols` already implement `PartialEq`